include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1583 @@
+      enable_taping(const types::tape_index,
+                    const bool,
+                    const types::tape_buffer_sizes,
+                    const types::tape_buffer_sizes,
+                    const types::tape_buffer_sizes,
+                    const types::tape_buffer_sizes)
+      {
+        AssertThrow(false, ExcRequiresAdolC());
+      }
+
+      static void
+      disable_taping(const types::tape_index, const bool)
+      {
+        AssertThrow(false, ExcRequiresAdolC());
+      }
+
+      static void
+      print_tape_stats(std::ostream &, const types::tape_index)
+      {
+        AssertThrow(false, ExcRequiresAdolC());
+      }
+
+
+      // === Scalar drivers ===
+
+      static scalar_type
+      value(const types::tape_index, const std::vector<scalar_type> &)
+      {
+        AssertThrow(false, ExcRequiresAdolC());
+        return 0.0;
+      }
+
+      static void
+      gradient(const types::tape_index,
+               const std::vector<scalar_type> &,
+               Vector<scalar_type> &)
+      {
+        AssertThrow(false, ExcRequiresAdolC());
+      }
+
+      static void
+      hessian(const types::tape_index,
+              const std::vector<scalar_type> &,
+              FullMatrix<scalar_type> &)
+      {
+        AssertThrow(false, ExcRequiresAdolC());
+      }
+
+      // === Vector drivers ===
+
+      static void
+      values(const types::tape_index,
+             const unsigned int,
+             const std::vector<scalar_type> &,
+             Vector<scalar_type> &)
+      {
+        AssertThrow(false, ExcRequiresAdolC());
+      }
+
+      static void
+      jacobian(const types::tape_index,
+               const unsigned int,
+               const std::vector<scalar_type> &,
+               FullMatrix<scalar_type> &)
+      {
+        AssertThrow(false, ExcRequiresAdolC());
+      }
+    };
+
+#  endif
+
+
+    // Specialization for ADOL-C taped numbers. It is expected that the
+    // scalar return type for this class is a float.
+    //
+    // Note: ADOL-C only has drivers for doubles, and so floats are
+    // not intrinsically supported. This wrapper struct works around
+    // the issue when necessary.
+    template <typename ADNumberType>
+    struct TapedDrivers<
+      ADNumberType,
+      float,
+      typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                              NumberTypes::adolc_taped>::type>
+    {
+      using scalar_type = float;
+
+      static std::vector<double>
+      vector_float_to_double(const std::vector<float> &in)
+      {
+        std::vector<double> out(in.size());
+        std::copy(in.begin(), in.end(), out.begin());
+        return out;
+      }
+
+      // === Taping ===
+
+      static void
+      enable_taping(const types::tape_index tape_index,
+                    const bool              keep_independent_values)
+      {
+        TapedDrivers<ADNumberType, double>::enable_taping(
+          tape_index, keep_independent_values);
```

Comment:
maybe add the comment `// ADOL-C only supports 'double', not 'float', so we can forward to the 'double' implementation of this function`

same below